### DISCUSSION TITLE
Old processes aren't killed when changes are detected while running in Docker

### DISCUSSION THREAD

**Discussion by binaryfire** [spatie/laravel-horizon-watcher#31] [Q&A]
Hi guys

I'm having issues when running `horizon:watch` in Docker. The package is detecting changes correctly, but it's starting new processes without killing the old ones. Which means I end up with an additional set of processes every time I save a file. It only takes a few saves before all my memory is used up.

I'm using Octane with `--watch` as well (in a different container) if that makes a difference.

**Comment by mihai-waydev** [comment]
Any news here? I also run it like this in the docker entrypoint:

`tini -s -- php artisan horizon:watch --without-tty & php artisan octane:start --host=0.0.0.0 --port=8083 --watch & wait -n`

but after a few code changes docker just crashes completely because it probably runs out of resources.

LE: Thanks for the response, binaryfire

**Comment by binaryfire** [comment]
@mihai-waydev I tried a bunch of ways, including using Supervisor + a file watcher that ran `horizon:terminate` on file changes. That kind of worked but still wasn't 100% reliable. We ended up switching back to a simple Redis queue with `queue:listen` for local development. Trying to run Horizon in local development just isn't worth it IMHO.

**Reply by AidasK** [reply]
Same here, too bad this issue is ignored and closed.

**Comment by gilney-canaltelecom** [comment]
I'm using Windows 11 + WSL2 with Debian + Docker with Ubuntu 22.04 image + Supervisor.

This i my supervisor entry:
```
[program:horizon_dev]
command=/usr/bin/php /var/www/html/artisan horizon:watch --without-tty
user=root
environment=LARAVEL_SAIL="1"
autostart=true
autorestart=true
stopasgroup=true
stopwaitsecs=3600
stdout_logfile=/var/log/supervisor/horizon_stdout.log
stdout_logfile_maxbytes=5MB
stderr_logfile=/var/log/supervisor/horizon_stderr.log
stderr_logfile_maxbytes=5MB
```

This is what happens when docker stats:
![1_Captura de tela 2025-01-07 144930](https://github.com/user-attachments/assets/a00484cc-8205-4171-9299-6c7e425aaf35)

And this is what happens after i change some file inside the `app/` folder:
![2_Captura de tela 2025-01-07 145131](https://github.com/user-attachments/assets/c47a315f-8b41-40ab-9b87-94400c9e1450)

You can't see it clearly, but the `horizon:watch` (1689) process is running inside the supervisor process and after the file change, the `php artisan horizon` (1697) process, that was inside the `horizon:watch`, is running outside it and a new `php artisan horizon` (1880) process is started.

If found [this thread](https://github.com/symfony/symfony/issues/20259) that seems to be related to the problem. 

If i make this change in this package' command directly inside `vendor/`, everything works correctly, the `php artisan horizon` command is really killed and a new one is started:

```diff
-$this->horizonProcess = Process::fromShellCommandline(config('horizon-watcher.command'))
-    ->setTty(!$this->option('without-tty'))
-    ->setTimeout(null);
+$this->horizonProcess = new Process(['php', 'artisan', 'horizon']);
+$this->horizonProcess->setTty(!$this->option('without-tty'))
+    ->setTimeout(null);
``` 

After the changes mentioned above, this is how the processes appear when docker starts and also after i edit my files:
![3_Captura de tela 2025-01-07 153238](https://github.com/user-attachments/assets/1756dcbc-cb63-4f64-8c9c-4fd76e36b74d)

As you can see, there is no `sh -c php artisan horizon` anymore.

I dont know if this causes other problems, but for now, for me at least, it works how i need it to work, so its fine 😅 Good luck guys.

*Edit:* So... I found other problems:
* Using the array version to pass the command to `Process` makes so the `php artisan horizon` output isn't processed in realtime;
* [The `Watch` lib throws an exception when the supervisor is restarted/stopped;](https://github.com/spatie/laravel-horizon-watcher/discussions/37)
* [The `onAnyChange` may be called multiple times if more than 1 file is edited at the same time, like with git;](https://github.com/spatie/laravel-horizon-watcher/discussions/24)

Here is the code that i edited to "fix" all these problems:

<details>
    <summary>Click me</summary>

```php
namespace App\Console\Commands;

use Illuminate\Console\Command;
use Spatie\Watcher\Exceptions\CouldNotStartWatcher;
use Spatie\Watcher\Watch;
use Symfony\Component\Process\Process;

class WatchHorizon extends Command
{
    protected Process $horizonProcess;

    protected $signature = 'horizon:watch {--without-tty : Disable output to TTY}';

    protected $description = 'Run Horizon and restart it when PHP files are changed';

    protected $trappedSignal = null;
    protected $didAlreadyRestart = false;

    public function handle()
    {
        $this->components->info('Starting Horizon, and restart when any file changes...');

        if (!$this->startHorizon()) {
            return Command::FAILURE;
        }

        $this->listenForChanges();
    }

    protected function startHorizon(): bool
    {
        $this->horizonProcess = new Process(explode(' ', config('horizon-watcher.command', '')));
        $this->horizonProcess->setTty(!$this->option('without-tty'))
            ->setTimeout(null);

        $this->trap([SIGINT, SIGTERM, SIGQUIT], function ($signal): void {
            $this->trappedSignal = $signal;

            // Forward signal to Horizon process.
            $this->horizonProcess->stop(signal: $signal);
            $this->horizonProcess->wait();
        });

        $this->horizonProcess->start(function (string $type, string $output): void {
            // I dont want the \n of '->info'/'->error'
            $type === 'out' ? $this->output->write($output) : $this->output->write("<error>{$output}</error>");
        });

        sleep(1);

        return !$this->horizonProcess->isTerminated();
    }

    protected function listenForChanges(): self
    {
        try {
            Watch::paths(config('horizon-watcher.paths'))
                ->onAnyChange(function (string $event, string $path) {
                    if ($this->changedPathShouldRestartHorizon($path)) {
                        $this->restartHorizon();

                        // 'debounce', since this function is called for EACH changed file
                        $this->didAlreadyRestart = true;
                    }
                })
                ->shouldContinue(function () {
                    // Necessary to get the horizon output in "realtime"
                    // It will be printed by the callback in '->start()'
                    $this->horizonProcess->getIncrementalOutput();
                    $this->horizonProcess->getIncrementalErrorOutput();

                    // 'debounce', this is called after all the changes are dealt with
                    $this->didAlreadyRestart = false;

                    return $this->trappedSignal === null;
                })
                ->start();
        } catch (CouldNotStartWatcher $e) {
            // Node process is killed before this one, so the watcher thinks it is not working
            // and throws this exception
            $this->error($e->getMessage());
        }

        return $this;
    }

    protected function changedPathShouldRestartHorizon(string $path): bool
    {
        if ($this->isPhpFile($path)) {
            return true;
        }

        foreach (config('horizon-watcher.paths') as $configuredPath) {
            if ($path === $configuredPath) {
                return true;
            }
        }

        return false;
    }

    protected function restartHorizon(): self
    {
        $this->components->info('Change detected! Restarting horizon...');

        $this->horizonProcess->stop();
        $this->horizonProcess->wait();

        $this->startHorizon();

        return $this;
    }

    protected function isPhpFile(string $path): bool
    {
        return str_ends_with(strtolower($path), '.php');
    }
}
```
</details>

PS: i didn't test everything very well, i just want to make this work for my dev setup for now, so dont use it blindly 😅

Its a real shame that the mantainers ignored all these problems till now, but i guess thats what happens when you have so many packages... ;/
